lib/logging/lib/logging.c

```diff
@@ -109,29 +109,29 @@ log_impl(
 
         merr_strerror(err, buf, sizeof(buf));
 
-        rc = snprintf(scratch_buffer, sizeof(scratch_buffer), "[HSE] %s:%d %s: %s: %s",
-            ev->ev_file, ev->ev_line, ev->ev_dte.dte_func, fmt, buf);
+        rc = snprintf(scratch_buffer, sizeof(scratch_buffer), "[HSE] %s:%d %s: %s: %s%s",
+            ev->ev_file, ev->ev_line, ev->ev_dte.dte_func, fmt, buf, log_file ? "\n" : "");
     } else {
-        rc = snprintf(scratch_buffer, sizeof(scratch_buffer), "[HSE] %s:%d %s: %s",
-            ev->ev_file, ev->ev_line, ev->ev_dte.dte_func, fmt);
+        rc = snprintf(scratch_buffer, sizeof(scratch_buffer), "[HSE] %s:%d %s: %s%s",
+            ev->ev_file, ev->ev_line, ev->ev_dte.dte_func, fmt, log_file ? "\n" : "");
     }
     if (rc >= sizeof(scratch_buffer)) {
         if (log_file) {
-            fprintf(log_file, "[HSE] %s:%d %s: scatch buffer size too small, needed %d for \"%s\"",
-                SRC_FILE, __LINE__, __func__, rc, scratch_buffer);
+            fprintf(log_file, "[HSE] %s:%d %s: scatch buffer size too small, needed %d for %s:%d\n",
+                SRC_FILE, __LINE__, __func__, rc, ev->ev_file, ev->ev_line);
         } else {
-            syslog(LOG_ERR, "[HSE] %s:%d %s: scatch buffer size too small, needed %d for \"%s\"",
-                SRC_FILE, __LINE__, __func__, rc, scratch_buffer);
+            syslog(LOG_ERR, "[HSE] %s:%d %s: scatch buffer size too small, needed %d for %s:%d",
+                SRC_FILE, __LINE__, __func__, rc, ev->ev_file, ev->ev_line);
         }
 
         return;
     } else if (rc < 0) {
         if (log_file) {
-            fprintf(log_file, "[HSE] %s:%d %s: bad printf format string: %s",
-                SRC_FILE, __LINE__, __func__, fmt);
+            fprintf(log_file, "[HSE] %s:%d %s: bad printf format string from %s:%d\n",
+                SRC_FILE, __LINE__, __func__, ev->ev_file, ev->ev_line);
         } else {
-            syslog(LOG_ERR, "[HSE] %s:%d %s: bad printf format string: %s",
-                SRC_FILE, __LINE__, __func__, fmt);
+            syslog(LOG_ERR, "[HSE] %s:%d %s: bad printf format string from %s:%d",
+                SRC_FILE, __LINE__, __func__, ev->ev_file, ev->ev_line);
```

Comment:
You could create function such as `log_xxx(FILE *log_file, int log_level, char *format, ...)` and have it call out to `vfprintf` or `syslogv` based on `log_file==NULL`.   It would be a nice cleanup / code reduction.

Reply:
you are right. Let me check that out.